### ISSUE TITLE
OpenIdConnect.nonce cookie lost on redirecting back to client from IdentityServer on Microsoft.Owin 3.1.0

### ISSUE THREAD

**Issue by ahokkonen-dev** [aspnet/AspNetKatana#70]
Hi,

we having some issues in production with missing authentication cookies after upgrading Microsoft.Owin.* libraries to 3.1.0 (been on 3.0.1) even with using different CookieManager provided by assembly.

The problem detected only on several (lets say 10 of 100) machines using FireFox/IE/Edge -browsers, but never on Chrome. During debug we see that OpenIdConnect.nonce cookie is setted well on client to Responce Cookies before redirecting to IdentityServer, but after successful login it is lost while redirected back to client - no OpenIdConnect.nonce found in Request and the infinite loop between app and IS as a result. As it was quite rare issue we didn't noticed it during test sessions.

We tried all kind of CookieManagers  (SystemWeb*, SystemWebChunking* etc) with no luck, so this situation forced us to downgrade back to 3.0.1 and it helped. 

Is anyone else have this kind of issue with 3.1.0?

**Comment by YovavGad** [comment]
I'm experiencing some issue too, but not exactly sure if it is related or not, on some operating systems (Linux, Linux X11, iOS) also Chrome, currently using Microsoft.Owin.3.1.0 v3.1.604505.82

This is weird because I'm only getting this on some OS/devices:

> The required anti-forgery cookie "__RequestVerificationToken" is not present.

> Void ValidateTokens(System.Web.HttpContextBase, System.Security.Principal.IIdentity, System.Web.Helpers.AntiXsrf.AntiForgeryToken, System.Web.Helpers.AntiXsrf.AntiForgeryToken)

Stack Trace

> at System.Web.Helpers.AntiXsrf.TokenValidator.ValidateTokens(HttpContextBase httpContext, IIdentity identity, AntiForgeryToken sessionToken, AntiForgeryToken fieldToken)
> at System.Web.Helpers.AntiXsrf.AntiForgeryWorker.Validate(HttpContextBase httpContext)
> at System.Web.Mvc.ControllerActionInvoker.InvokeAuthorizationFilters(ControllerContext controllerContext, IList`1 filters, ActionDescriptor actionDescriptor)
> at System.Web.Mvc.Async.AsyncControllerActionInvoker.<>c__DisplayClass21.b__19(AsyncCallback asyncCallback, Object asyncState)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.Begin(AsyncCallback callback, Object state, Int32 timeout)
> at System.Web.Mvc.Async.AsyncControllerActionInvoker.BeginInvokeAction(ControllerContext controllerContext, String actionName, AsyncCallback callback, Object state)
> at System.Web.Mvc.Controller.b__1c(AsyncCallback asyncCallback, Object asyncState, ExecuteCoreState innerState)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncVoid`1.CallBeginDelegate(AsyncCallback callback, Object callbackState)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.Begin(AsyncCallback callback, Object state, Int32 timeout)
> at System.Web.Mvc.Controller.BeginExecuteCore(AsyncCallback callback, Object state)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.Begin(AsyncCallback callback, Object state, Int32 timeout)
> at System.Web.Mvc.Controller.BeginExecute(RequestContext requestContext, AsyncCallback callback, Object state)
> at System.Web.Mvc.MvcHandler.b__4(AsyncCallback asyncCallback, Object asyncState, ProcessRequestState innerState)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncVoid`1.CallBeginDelegate(AsyncCallback callback, Object callbackState)
> at System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResultBase`1.Begin(AsyncCallback callback, Object state, Int32 timeout)
> at System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContextBase httpContext, AsyncCallback callback, Object state)
> at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
> at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
> 
>

**Comment by ahokkonen-dev** [comment]
No sure it is directly related to our problem, but it may.
In our situation the exception is standard, thrown from OpenIdConnectAuthenticationNotifications AuthenticationFailed -method:

> IDX10311: RequireNonce is 'true' (default) but validationContext.Nonce is null. A nonce cannot be validated. If you don't need to check the nonce, set OpenIdConnectProtocolValidator.RequireNonce to 'false'.

Some kind of nonce overflow usually related to IDX10311 error, but not in our case.

**Comment by Meyce** [comment]
We've been encountering an issue with IDX10311, except for us a set-cookie is not being issued with the nonce cookie and upon successful login at the IDP and redirect back to client, there is no nonce cookie resulting in IDX10311.  Also, this does not consistently happen for us, making it more difficult to pin down.

**Comment by Meyce** [comment]
@ahokkonen-dev, have you checked to see if the nonce cookie is expired?  I just checked on 3.1.0 the expiration on the cookie is 15 minutes, so if you sat at the login for more than 15 minutes you could get IDX10311.

**Comment by ahokkonen-dev** [comment]
@Meyce , cookie expiration is not what causing it in our case. Tried to login right after we landed on IDP login page and on redirect back nonce cookie is already missing. 

As in your case this happens on rare occasion and we completely out of ideas which direction we need to dig in :)

**Comment by Tratcher** [comment]
Can you share a Fiddler trace of your failing flow?

**Comment by michaelharrod** [comment]
We've been able to reliably recreate a missing nonce error "IDX10311: RequireNonce is 'true' (default) but validationContext.Nonce is null." This is what we do:

1. Open IE11, browse to your site (RP) that uses Azure AD (IdP) authentication.
2. Be redirected to IdP, enter username and password as usual.
3. Be redirected to RP notice normal login.
4. Sign out of RP.
5. Open Internet Options
6. Ensure Delete browsing history on exit option selected.
7. Click Delete...
8. Ensure Preserve Favourites website data NOT ticked.
9. Ensure Cookies and website data ticked.
10. Click Delete.
11. Click OK
12. Close the browser
13. Reopen the browser and browse to the RP.
14. Be redirected to IdP, enter username and password as usual.
15. Notice error.

Looking at the Fiddler traffic we can clearly see that the RP sets the nonce cookie, but when the browser returns to the RP after authenticating, the cookie is not sent. The cookie's expiry date looks fine.

Can anyone replicate the error by following the above steps?

**Comment by ahokkonen-dev** [comment]
Miracly we managed to fix this issue by adding empy Session_Start -method to Global.asax as suggested at this thread: [https://github.com/IdentityServer/IdentityServer3/issues/542](url). After that no more missing cookies or soever :)

I really don't understund how it works but it really do!

**Comment by michaelharrod** [comment]
We had the infinite redirect issue originally and this SystemWebCookieManagerfixed it for us:
https://stackoverflow.com/a/35172649/1825109

This new missing nonce error looks like a different cause. Strangely a downstream result of clearing browser history in IE 11 causing it to fail to retain a new cookie that's presented to it. Our tester only raised this because we kept telling him to clear his cookies when he reported other bugs.

**Comment by ahokkonen-dev** [comment]
Oh boy, now we managed to find out what was causing this issue!
TimeZone was set wrong on those machines and that was it. Basically on those machines where timezone was "behind" current time on creating OpenIdConnect.nonce cookie it was already expired and cleaned up right after redirection to identity provider. 

It was not an issue for 3.0.1 but on 3.1.0 it is time zone sensitive!

**Comment by YJChan** [comment]
Hi, We happened to face the same error,  missing nonce error "IDX10311: RequireNonce is 'true' (default)". Most of the cases, we are failed to login.  We are following the guide provided by Microsoft. 
https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-v1-aspnet-webapp

After some finding, we have try to capture all our http headers with fiddler and realized that a cookie will be set, named OpenIdConnect.nonce.6kgBFpe7Jf..... at my path. If this cookie does not exist, the login is always failed. There is no hint on how to properly solved this thing. 

However, we applied @ahokkonen-dev  method. After adding empty Session_OnStart() method in our global.asax this error is not exist anymore. Not sure what is the reasons behind but it works.
